### ISSUE TITLE
The regex that detects mentions is assuming English (and similar) languages

### ISSUE THREAD

**Issue by yossizahn** [julianlam/nodebb-plugin-mentions#121]
The regex that detects "mentions" is assuming that an `@` character is only the start of a mention if it follows a whitespace character (or a very limited number of other characters). 
This assumption isn't true in other languages where a preposition may be expressed by prefixing the subject word.
e.g. In English we say "... that @XYZ said..."
In Hebrew this is written as: "ש-@XYZ אמר", or even as "ש@XYZ אמר" without the hyphen.

**Comment by julianlam** [comment]
Understood, but unfortunately I had to make concessions as to how the mentions regex matcher works...

For example, I don't want it to trigger the mentions autocomplete dialog when you simply type an email address (e.g. whatever@example.com)

Unfortunately, I can't think of a solution around this. Do other services provide mentions in this manner?

I am attempting to try with GitHub right now, but it also seems to want a whitespace character before the mention.

**Comment by yossizahn** [comment]
> For example, I don't want it to trigger the mentions autocomplete dialog when you simply type an email address 

Perhaps detect and linkify mentions anywhere they occur but only trigger the autocomplete when after a whitespace. 

> I am attempting to try with GitHub right now, but it also seems to want a whitespace character before the mention.

Github seems to be quite clever about this and only linkifies mentions which are not proceeded by [0-9a-zA-Z]
It is also has stricter rules for triggering autocomplete than for linkifying.

**Comment by julianlam** [comment]
Right... and just like GitHub (not that I'm comparing or anything), we also have separate logic for linkifying mentions vs. spawning the autocomplete dialog.

They each have their own regex, although they are for all intents and purposes identical when it comes to matching a mention.

It's because of unicode characters that we couldn't simply check for the word boundary meta character `\b`... we could use the XRegExp library for server-side (so... the linkifying part), but it wouldn't resolve the autocomplete issue. If that's an okay concession for you, then I can look into maybe implementing XRegExp (though no guarantees)

**Comment by yossizahn** [comment]
> If that's an okay concession for you,

Anything is better than nothing :)
The main issue is actually not the auto-complete or the linkifying but the fact that the user doesn't get notified about the mention.
Fixing it server side is enough for that.
